Use `CAMLnoret` instead of `CAMLnoreturn_{start,end}`

xavierleroy opened the pull request:
`_Noreturn` is standard C11, shorter and more visually pleasant.  Moreover, it can be put directly on functions definitions, unlike the `CAMLnoreturn_{start,end}` macros, which apply to function declarations only.

The `CAMLnoreturn_{start,end}` and  `Noreturn` macros remain defined, this should not break any 3rd-party code.

xavierleroy committed:
Use `_Noreturn` instead of `CAMLnoreturn_{start,end}`

`_Noreturn` is standard C11, shorter, more visually pleasant,
and it can be put directly on functions definitions.

The `CAMLnoreturn_{start,end}` and  `Noreturn` macros remain defined,
this should not break any 3rd-party code.

xavierleroy commented:
I considered adding a deprecation warning for `Noreturn` (the old macro that didn't work with MSVC) and possibly for `CAMLnoreturn_{start,end}` (the more recent macros that work both with GCC/Clang and MSVC), but I'm unsure we really need to bother users with this.

MisterDA commented:
In C23, the `_Noreturn` function specifier is deprecated. [`[[noreturn]]`](https://en.cppreference.com/w/c/language/attributes/noreturn) attribute should be used instead. The macro `noreturn` is also deprecated.
I'm not fond of adding yet another `CAMLnoreturn` macro, yet we might consider already supporting upcoming standards. Something like

```c
#if __STDC_VERSION__ == 201112L || __STDC_VERSION__ == 201710L
#define CAMLnoreturn _Noreturn
#elif __STDC_VERSION__ >= 201710L || __cplusplus >= 201103L
#define CAMLnoreturn [[noreturn]]
#else
#define CAMLnoreturn
#endif
```

xavierleroy commented:
> In C23, the _Noreturn function specifier is deprecated. [[[noreturn]]](https://en.cppreference.com/w/c/language/attributes/noreturn) attribute should be used instead. The macro noreturn is also deprecated.

C is moving so fast these days!  Deprecating a feature that was introduced 12 years ago only!  We would never do that in OCaml :-)

More seriously: I don't expect OCaml to require C23 any time soon, if only because we cannot get rid of the MSVC compiler, which tends to follow C standards with a 10+ year delay.

However, your message reminds me of C++ compatibility for OCaml's headers.  Am I correct that `_Noreturn` is not in any C++ standard?  And that it's `[[noreturn]]` or GCC-style attributes?

MisterDA commented:
> > In C23, the _Noreturn function specifier is deprecated. [[[noreturn]]](https://en.cppreference.com/w/c/language/attributes/noreturn) attribute should be used instead. The macro noreturn is also deprecated.
> 
> C is moving so fast these days! Deprecating a feature that was introduced 12 years ago only! We would never do that in OCaml :-)

They maybe counting on the fact that no one has adopted the too new feature in 12 years, and leap to the new standard? :-)

> More seriously: I don't expect OCaml to require C23 any time soon, if only because we cannot get rid of the MSVC compiler, which tends to follow C standards with a 10+ year delay.

Oh, I forgot about this one. On the other hand, when other compilers switch to the new standard by default, they often introduce new warnings against deprecated features too.

> However, your message reminds me of C++ compatibility for OCaml's headers. Am I correct that `_Noreturn` is not in any C++ standard? And that it's `[[noreturn]]` or GCC-style attributes?

That seems to be the case, since C++11. At least it's what [cppreference](https://en.cppreference.com/w/cpp/language/attributes/noreturn) shows.

xavierleroy commented:
All right, something needs to be done, then.  Sadly, `CAMLnoreturn` is already defined in `<caml/memory.h>` with a different meaning...  Should we use `CAMLnoret`?  Or repurpose the existing `Noreturn` macro?

xavierleroy committed:
Use `CAMLnoret` instead of `CAMLnoreturn_{start,end}`

`CAMLnoret` is a new macro in `<caml/misc.h>` that can be put at the beginning
of function declarations or function definitions, and expands to whatever
is most standard and supported by the C / C++ compiler.

xavierleroy committed:
yacc/defs.h: use CAMLnoret instead of Noreturn

xavierleroy commented:
I pushed an attempt along the lines above.  I'm not super happy about this code: too many `#if`, and in the end the most portable approach (GCC's attributes) is also the least standardized...  So, unless there is clear interest, I'll probably close this PR.

MisterDA commented:
What if, as a temporary measure, we keep on using `CAMLnoreturn_start`, but the above "modern" definitions? If I'm not mistaken that would also make the noreturn attribute work with more compilers than just GNUC and msvc. 
Otherwise it seems that the macros definitions are quite unavoidable. They might be simplified if we required a complete C11/C++ compiler, because the `_Noreturn` keyword, resp the `[[noreturn]]` attribute are always available. 
I did not think through what changing `CAMLnoreturn_start` would imply for function declarations/definitions.

Having the noreturn macro available is useful for the clang static analyzer which I use from time to time on ocaml and my C stubs. It suppresses false-positive warnings. I suppose it also helps the analyzer in other ways I don't know about.

xavierleroy commented:
> What if, as a temporary measure, we keep on using CAMLnoreturn_start, but the above "modern" definitions? If I'm not mistaken that would also make the noreturn attribute work with more compilers than just GNUC and msvc.

I was really hoping to make our code and users' code more readable than it is today, avoiding two shortcomings of the `CAMLnoreturn_start`/`CAMLnoreturn_end` macros: 1- they must occur in matching pairs and 2- they don't apply to function definitions, only function declarations.

The GNU C `attribute((noreturn))` is supported by a great many compilers, from Clang to CompCert through ICC, so portability is not a major concern here.

MisterDA commented:
That makes sense. If you're not convinced by your PR as-is and don't see how to simplify it further, then it's probably better to drop it, but it's not worse than what we currently have, so altogether I'm in favor of merging.

gasche reviewed (approved):
I agree with @MisterDA that the PR improves things -- the macro definition is letter soup as before, but the use-sites are lighter and nicer to read now, especially those that had to have both a declaration and a definition.

dra27 commented on runtime/caml/misc.h:
```diff
@@ -75,28 +75,32 @@ typedef size_t asize_t;
+  #define CAMLnoret _Noreturn
+#elif defined(__GNUC__)
+  #define CAMLnoret  __attribute__ ((noreturn))
+#elif defined(_MSC_VER)
+  #define CAMLnoret __declspec(noreturn)
```
MSVC will definitely return with `/std:c11`, which correctly supports `_Noreturn` above (and defines `__STDC_VERSION__` correct, too)

xavierleroy replied:
OK, then I'm happy to remove this special case.

dra27 commented on runtime/caml/misc.h:
```diff
@@ -75,28 +75,32 @@ typedef size_t asize_t;
+  #define CAMLnoret [[noreturn]]
+#elif __STDC_VERSION__ >= 201112L
+  #define CAMLnoret _Noreturn
+#elif defined(__GNUC__)
+  #define CAMLnoret  __attribute__ ((noreturn))
```
```suggestion
#elif defined(__GNUC__) /* Header processed in C++ mode */
  #define CAMLnoret  __attribute__ ((noreturn))
```
or words to that effect. I got puzzled reading it, and I could imagine it being accidentally hoovered in future!

xavierleroy replied:
No, C++ is not the (main) reason for this special case.  Some not too old, not too recent versions of GCC support all of C11 but default to gnu99 mode, so `__STDC_VERSION__` corresponds to C99 and the macros above don't select `_Noreturn`.

dra27 reviewed (approved):
I agree that the neater call sites are reason enough for the change! Two tiny tweaks in misc.h, if you agree, but otherwise let's merge?

xavierleroy committed:
No need to special-case pre-C11 MSVC

xavierleroy committed:
Changes for #12235

xavierleroy commented:
This PR triggers a warning introduced in GCC 12.  It's a long story that I'll tell in another PR along with a workaround.  In the meantime, this PR should not be merged.

xavierleroy commented:
#12341 tells the long story and fixes the spurious warning, so let's merge this PR.
